client/me/get-apps/mobile-download-card.jsx

```diff
@@ -6,46 +6,263 @@
+		}
+
+		// Account recovery number already has the keys formatted in the
+		// way we want, so we can just return it directly.
+		if ( account_recovery_number !== null ) {
+			const isValid = this.phoneNumberIsValid( account_recovery_number.numberFull );
+			account_recovery_number.isValid = isValid;
+
+			return account_recovery_number;
+		}
+
+		// Fallback if we didn't match anything
+		return noPreferredNumber;
+	};
+
+	phoneNumberIsValid( number ) {
+		return ! phoneValidation( number ).error;
+	}
+
+	numericCountryCodeForCountryCode( code ) {
+		const element = this.props.countriesList.find( item => {
+			return item.code === code;
+		} );
+
+		if ( element !== undefined ) {
+			return element.numeric_code;
+		}
+
+		return null;
+	}
+
+	render() {
+		const translate = this.props.translate;
+
+		const has_all_data = this.props.hasUserSettings && this.props.hasLoadedAccountRecoveryPhone;
+		const { countryCode, number, isValid } = this.getPreferredNumber( has_all_data );
+		const { isMobile } = userAgent;
+		const feature_is_enabled = config.isEnabled( 'get-apps-sms' ) && ! isMobile;
+
+		return (
+			<Card className="get-apps__mobile">
+				<div className="get-apps__store-subpanel">
+					<div className="get-apps__card-text">
+						<h3 className="get-apps__card-title">{ translate( 'Mobile Apps' ) }</h3>
+						<p className="get-apps__description">
+							{ translate( 'WordPress at your fingertips.' ) }
+						</p>
+					</div>
+					<div className="get-apps__badges">
+						<AppsBadge
+							storeLink="https://play.google.com/store/apps/details?id=org.wordpress.android&referrer=utm_source%3Dcalypso-get-apps%26utm_medium%3Dweb%26utm_campaign%3Dmobile-download-promo-pages"
+							storeName={ 'android' }
+							titleText={ translate( 'Download the WordPress Android mobile app.' ) }
+							altText={ translate( 'Google Play Store download badge' ) }
+						/>
+						<AppsBadge
+							storeLink="https://itunes.apple.com/app/apple-store/id335703880?pt=299112&ct=calpyso-get-apps-button&mt=8"
+							storeName={ 'ios' }
+							titleText={ translate( 'Download the WordPress iOS mobile app.' ) }
+							altText={ translate( 'Apple App Store download badge' ) }
+						/>
+					</div>
+				</div>
+
+				{ feature_is_enabled ? (
+					<div className="get-apps__sms-subpanel">
+						<p>
+							<strong>{ translate( 'Ready to WordPress on the go?' ) }</strong>
+							<br />
+							{ translate(
+								'We’ll send you an SMS message with a download link for the right app for your mobile device.'
+							) }
+						</p>
+
+						<div className="get-apps__sms-field-wrapper">
+							{ has_all_data ? (
+								<FormPhoneInput
+									countriesList={ this.props.countriesList }
+									initialCountryCode={ countryCode }
+									initialPhoneNumber={ number }
+									phoneInputProps={ {
+										onKeyUp: this.onKeyUp,
+									} }
+									onChange={ this.onChange }
+								/>
+							) : (
+								<>
+									<QuerySmsCountries />
```

Comment:
Don't hesitate to move to up in the tree, rendering them without conditions is the best practice I think.

Reply:
I moved the data fetching tags out of the conditional, but I need to leave the empty `<>` there – it's true that it's suboptimal to re-render the nodes, but if we don't, there's no way to set the user's phone number when the page loads. Unfortunately, `FormPhoneInput` only accepts an initial value, and can't be updated after that. Rather than refactoring it and re-working the code that's used in the rest of the codebase (or forking the component and causing more confusion) this seemed simpler.

It's a bit of a hack – without wrapping `FormPhoneInput` in the `<>`, when React diffs the DOM, it doesn't replace the component, so the phone number never populates. By having a different tree in the conditional, it re-creates the component from scratch, and properly populates the phone number.

LMK if you'd like approach it differently – this way seemed like the lowest-impact option :)